imports/wasi_snapshot_preview1/poll.go

```diff
@@ -65,95 +77,148 @@ func pollOneoffFn(ctx context.Context, mod api.Module, params []uint64) syscall.
-			outBuf[outOffset+8] = 0
+	// process timeout and interactive inputs (if any)
+	if timeout > 0 && readySubs == 0 {
+		timeoutCtx, cancelFunc := context.WithTimeout(ctx, timeout)
+		defer cancelFunc()
+
+		for _, s := range stdinSubs {
+			go processDelayedStdinReader(s, outBuf, cancelFunc)
 		}
-		outBuf[outOffset+9] = 0
-		le.PutUint32(outBuf[outOffset+10:], uint32(eventType))
-		// TODO: When FD events are supported, write outOffset+16
+
+		<-timeoutCtx.Done()
 	}
+
 	return 0
 }
 
 // processClockEvent supports only relative name events, as that's what's used
 // to implement sleep in various compilers including Rust, Zig and TinyGo.
-func processClockEvent(_ context.Context, mod api.Module, inBuf []byte) syscall.Errno {
+func processClockEvent(inBuf []byte) (time.Duration, syscall.Errno) {
 	_ /* ID */ = le.Uint32(inBuf[0:8])          // See below
 	timeout := le.Uint64(inBuf[8:16])           // nanos if relative
 	_ /* precision */ = le.Uint64(inBuf[16:24]) // Unused
 	flags := le.Uint16(inBuf[24:32])
 
+	var err syscall.Errno
 	// subclockflags has only one flag defined:  subscription_clock_abstime
 	switch flags {
 	case 0: // relative time
 	case 1: // subscription_clock_abstime
-		return syscall.ENOTSUP
+		err = syscall.ENOTSUP
 	default: // subclockflags has only one flag defined.
-		return syscall.EINVAL
+		err = syscall.EINVAL
 	}
 
-	// https://linux.die.net/man/3/clock_settime says relative timers are
-	// unaffected. Since this function only supports relative timeout, we can
-	// skip name ID validation and use a single sleep function.
+	if err != 0 {
+		return 0, err
+	} else {
+		// https://linux.die.net/man/3/clock_settime says relative timers are
+		// unaffected. Since this function only supports relative timeout, we can
+		// skip name ID validation and use a single sleep function.
 
-	sysCtx := mod.(*wasm.ModuleInstance).Sys
-	sysCtx.Nanosleep(int64(timeout))
-	return 0
+		return time.Duration(timeout), 0
+	}
 }
 
-// processFDEvent returns a validation error or syscall.ENOTSUP as file or socket
-// subscriptions are not yet supported.
-func processFDEvent(mod api.Module, eventType byte, inBuf []byte) syscall.Errno {
-	fd := le.Uint32(inBuf)
-	fsc := mod.(*wasm.ModuleInstance).Sys.FS()
-
-	// Choose the best error, which falls back to unsupported, until we support
-	// files.
-	errno := syscall.ENOTSUP
-	if eventType == wasip1.EventTypeFdRead {
+func processFDEvent(fsc *internalsys.FSContext, fd uint32, e *event) *internalsys.StdioFileReader {
+	if e.eventType == wasip1.EventTypeFdRead {
 		if f, ok := fsc.LookupFile(fd); ok {
-			st, _ := f.Stat()
-			// if fd is a pipe, then it is not a char device (a tty)
-			if st.Mode&fs.ModeCharDevice != 0 {
-				errno = syscall.EBADF
+			// if fd corresponds to an interactive StdioFileReader then return it
+			if reader, ok := f.File.(*internalsys.StdioFileReader); ok && reader.IsInteractive() {
+				return reader
+			} else {
+				e.errno = wasip1.ErrnoSuccess
```

Comment:
Stdin should be checked here as well otherwise `select` can report stdin even when there is no input happens. 

```c
#include <stdio.h>
#include <unistd.h>
#include <sys/select.h>

int main(int argc, char **argv)
{
  char buf[5];
  int len = 5, ret;
  fd_set rfds;
  struct timeval tv;

  FD_ZERO(&rfds);
  FD_SET(0, &rfds);

  tv.tv_sec = 3;
  tv.tv_usec = 0;
  printf("waiting for inputs...\n");
  ret = select(1, &rfds, NULL, NULL, &tv);
  if ((ret > 0) && FD_ISSET(0, &rfds)) {
    printf("input detected (ret=%d), reading...\n", ret);
    ret = read(0, buf, len);
    if (ret > 0) {
      printf("got %.*s (first 5 chars)\n", ret, buf);
    }
  } else {
    printf("no input\n");
  }
}
```

host:

```
$ cat | ./a.out
waiting for inputs...
no input
```

On wazero, it reports inputs even when no input happens and the following `read` blocks.

```
$ cat | go run main.go ./out/out.wasm
waiting for inputs...
input detected (ret=1), reading...
```

Reply:
thanks!